Fixes #15286 - New API for auth sources

dhlavac opened the pull request:
Basic concept of new auth sources API

Commit:
Fixes #19039 - Lock plugin roles

dhlavac committed:
Fixes #15286 - New API for auth sources

lizagilman committed:
Fixes #20454 - change page by typing page number

lzap committed:
Fixes #20475 - Random DB IPAM with MAC preservation

Commit:
Fixes #20529 - Improve the filtering function

amirfefer committed:
Fixes #20416 - fix UI under email preferences

dLobatog committed:
Fixes #20543 - implement asset_path for plugin_assets

Sprockets runs on its own context in the task plugin_assets.rake since
 #20287. This means that the asset paths are not visible to it
immediately, the way we provide that functionality is by overriding a
method in Sprockets that gets the paths from the Rails app since we're
on Rake context.

ehelms committed:
Fixes #19529: Use main_app for redirect in case of plugins

Sometimes when entering the require_mail filter, the plugins context
is carried over and the route cannot be matched. Using main_app ensures
that it will resolve the path from the proper context.

yaunj committed:
Fixes #20537: Monospace fonts in fullscreen editor 

The fullscreen editor is used to edit parameters which can be in YAML.
Since YAML is an indentation sensitive format, monospaced fonts makes
for easier editing.

amirfefer committed:
Fixes #20255 - add csv export for hostgroups

avylove committed:
Fixes #20514 - Accept dollar sign($) in login names

dLobatog committed:
Refs #20514 - Add test for $ in user login

snagoor committed:
Fixes #20553 - br tag now renders to new line

lzap committed:
Fixes #20521 - better subnet prefix label

Commit:
Fixes #19326 - create compute-resource - URL optional

ohadlevy committed:
fixes #20316 - allow to delete a user that has notifications

tstrachota commented on app/controllers/api/v2/auth_source_controller.rb:
```diff
@@ -0,0 +1,27 @@
+        @auth_source = resource_scope_for_index
+      end
+
+      api :GET, "/auth_source/:id/", N_("Show all authentication sources")
```
"Show an authentication source"

tstrachota commented on app/controllers/api/v2/auth_source_external_controller.rb:
```diff
@@ -0,0 +1,26 @@
+      include Api::TaxonomyScope
+      before_action :find_resource, :only => %w{show}
+
+      api :GET, "/auth_source_external/", N_("List external audits")
```
"List external authentication sources"

tstrachota commented on app/controllers/api/v2/auth_source_internal_controller.rb:
```diff
@@ -0,0 +1,26 @@
+      include Api::TaxonomyScope
+      before_action :find_resource, :only => %w{show}
+
+      api :GET, "/auth_source_internal/", N_("List internal audits")
```
"List internal authentication sources"

tstrachota commented on app/controllers/api/v2/auth_source_external_controller.rb:
```diff
@@ -0,0 +1,26 @@
+      param_group :search_and_pagination, ::Api::V2::BaseController
+
+      def index
+        @auth_source_internal = resource_scope_for_index
```
Shouldn't this be `@auth_source_external`?

tstrachota commented on app/views/api/v2/auth_source/base.jsno.rabl:
```diff
@@ -0,0 +1,3 @@
+object @auth_source
```
There's a typo in the file name: `base.jsno.rabl` vs base.json.rabl

tstrachota commented on app/views/api/v2/auth_source/base.jsno.rabl:
```diff
@@ -0,0 +1,3 @@
+object @auth_source
+
+attributes :id, :type, :name
```
No newline at the end of file.

tstrachota commented on app/views/api/v2/auth_source_external/base.jsno.rabl:
```diff
@@ -0,0 +1,3 @@
+object @auth_sources_external
```
Ditto file name

tstrachota commented on app/views/api/v2/auth_source_external/test.json.rabl:
```diff
@@ -0,0 +1,10 @@
+end
+node :message do
+  locals[:message]
+end
```
Where is this `test.json.rabl` used?

tstrachota commented on app/views/api/v2/auth_source_external/create.json.rabl:
```diff
@@ -0,0 +1,3 @@
+object @auth_source_external
+
+extends "api/v2/auth_source_external/show"
```
There's no `create` action defined in the auth source external controller.

tstrachota commented on app/views/api/v2/auth_source_external/base.jsno.rabl:
```diff
@@ -0,0 +1,3 @@
+object @auth_sources_external
+
+attributes :id, :type, :name
```
It would be good to map `type` onto some human readable type name to avoid leaking the internal class names out of the api.

tstrachota commented on app/views/api/v2/auth_source/main.json.rabl:
```diff
@@ -0,0 +1,5 @@
+
+extends "api/v2/auth_source/base"
+
+attributes :onthefly_register, :usergroup_sync, :tls, :server_type, :created_at, :updated_at
```
I don't think it makes sense to display the fields above (apart from the timestamps) for a generic auth source. They're all LDAP related.

dhlavac replied:
I have found records of this attributes in database due to I think it make senes to show that. Which is related for all authsorces ? - :created_at, :updated_at ? Some else ?

tstrachota replied:
Yeah, it's caused by STI. It can be confusing.
From the list above I think it's only the timestamps.

Commit:
Fixes #19554 - Warning for unsupported PXE loader combinations

tstrachota commented on app/controllers/api/v2/auth_source_controller.rb:
```diff
@@ -0,0 +1,27 @@
+      api :GET, '/organizations/:organization_id/auth_source/',
+          N_('List authentication sources per organization')
+      param_group :taxonomy_scope, ::Api::V2::BaseController
+      param_group :search_and_pagination, ::Api::V2::BaseController
```
At the moment I don't think there's scoped search defined for auth sources. It would imho make sense to add it at least on name attr.

tstrachota reviewed (changes requested):
Thanks for the PR @dhlavac 

I added some comments into the code. Routes definition seem to be missing completely so it's not possible to use the new controllers.

Could you please add some tests for the controllers?

ares commented on app/views/api/v2/auth_source_internal/base.jsno.rabl:
```diff
@@ -0,0 +1,3 @@
+object @auth_sources_internal
+
+attributes :id, :name
```
missing new line at the end of the file

tbrisker committed:
Fixes #20610 - Prevent chrome password autofill

mmack committed:
Fixes #20608 - Adding index for ssh_keys

Adds a index :user_id to ssh_keys

amirfefer committed:
Fixes #20612 - Fix error rendering when building pxe default

amirfefer committed:
Fixes #20614 - fix host selector for bulk action regression

lzap committed:
Fixes #20622 - puppet import does not refresh all statuses

amirfefer committed:
Fixes #20628 - cloned host/hostgroup should copy rootpass

When cloning host/hostgroup, root pass field should be copied too

ohadlevy committed:
fixes #20636 - update babel-core and loader to newer versions

timogoebel committed:
fixes #20645 - stub templates proxy url call

timogoebel committed:
fixes #20646 - clear host_id during interface cloning

tbrisker committed:
Refs #20255 - simplify hostgroup csv test

ares committed:
Fixes #20662 - no failed features

bastilian committed:
Fixes #20573 - Include searchbox extension

bastilian committed:
Fixes #20669 - Append 'clone' to cloned provisioning templates

ares committed:
Fixes #20672 - add button for autosign entry

bastilian committed:
Fixes #20673 - Move locked warning on template editing

Moving it to the top above name to be more visible.

tbrisker committed:
Fixes #7499 - Allow searching config group by puppetclass

bastilian committed:
Fixes #20678 - Improve provisioning templates history

* Make glyphicons the default text color
* Add "No comment provided" when it is empty
* Change link actions to buttons
* Fix issue with empty tab when skipping
  non audit_templates

bastilian committed:
Fixes #20680 - Use default button style for "Add Combination"

ares committed:
Fixes #20676 - relabel new button for autosign entry

ares committed:
Fixes #20675 - autosign delete buttons are no longer red

ares committed:
Fixes #20668 - capitalize cards on smart proxy page

ares committed:
Fixes #20706 - fix template audit messages

ares committed:
Fixes #20710 - fix view diff width

snagoor committed:
Fixes #20711 - GET host interface API will show fqdn attribute

iNecas committed:
Fixes #19113 - use fog-ovirt method for rebooting

While fixing the waiting for stopped vm in reboot there.

ares committed:
Fixes #20726 - allow # in user login

mmoll committed:
Fixes #1644,#19692 - update fog-libvirt to 0.4.1

ares committed:
Fixes #7778 - allow + in user login

ares committed:
Fixes #20736 - improve bookmark permissions

kgaikwad committed:
Fixes #20701 - Subnet6 fields info in GET api/v2/hostgroups/:id

parthaa committed:
fixes #20699 - Rake task to selectively backup tables

This commit adds an optional param "tables" to db:dump task
One can use to this selectively dump data from desired tables.
Its main purpose is to curate debug data from customers for analysis.
For example one can run

rake db:dump RAILS_ENV=production tables="foreman_tasks.*, dynflow.*"
to get a copy of all the foreman-tasks and dynflow tables

Then on a local db one can choose to just restore those tables
and analyze

ares committed:
Fixes #20657 - make proxy labels bold

ares committed:
Fixes #20739 - use unique user label in select

ron-benaim committed:
Fixes #20750 - Comment field in host page

dLobatog committed:
i18n - extracting new, pulling from tx

dLobatog committed:
Bump version to 1.17-develop

ohadlevy committed:
fixes #20833 - allow storybook to use application css

bastilian committed:
Fixes #20763 - Reword "Create SSH key" to "Add SSH key"

bastilian committed:
Fixes #20801 - Put granularity info into alert box

amirfefer committed:
Fixes #20240 - fix fullscreen mode in ace editor

ShimShtein committed:
Fixes #13599 - Wrapped exception truncate moved to client

Removed message truncation from server side, and moved it to client
side - the message is truncated with ellipsis. On mouse over full
message is presented as popover.

This commit required a small change to `data-ajax-url` handler in client
to enable proper tooltip presentation in HTML loaded by AJAX.

ares committed:
Fixes #20737 - add default description to roles

lzap committed:
Fixes #20835 - added time into foreman-debug

timogoebel committed:
fixes #20820 - set ajax vars for cr host import

iNecas committed:
Fixes #20654 - color of Failed Restarts in resources chart

treydock committed:
Fixes #20500 - Allow GET /api/hostgroups to include parameters

Add include Array parameter to GET /api/hostgroups v2 that will
optionally include a hostgroup's parameters when the include parameter
contains 'parameters'.

bastilian committed:
Refs #19554 - Rewrite to use an Object and not Map

Map is not yet fully supported in all browsers and
would require additional dependencies to be added.

Using a Object with lodash functions achieves the
same and does not require any additional
dependencies.

kgaikwad committed:
Fixes #20858 - pxe_loader in GET api/v2/hostgroups/:id

timogoebel committed:
fixes #20822 - vm: show link to associated host

ares committed:
Fixes #20867 - pin down scoped search

snagoor committed:
Fixes #20516 - GET hosts API displays owner_name attribute

ares committed:
Fixes #20589 - update login_at when extlogin is used

dLobatog committed:
Fixes #20872 - scoped_search should be >= 4.1.2

As to introduce the fix for relations between classes with polymorphism
(https://github.com/wvanbergen/scoped_search/commit/e629a484cc57c65cd394e47006eb30b24e4382bc)

kgaikwad committed:
Fixes #20840 - show error on invalid name for any parameter

Updating any type of Parameter such as
subnet_parameter/host_parameter with invalid name
doesn't show any error.
With this commit, it is now showing error on parameters with
invalid name.

dLobatog committed:
Fixes #19730 - Add vendor to provisioning templates

Before unlocking a provisioning template, a warning is shown on the
screen. This warning should contain the vendor that provided the
template (normally Foreman or Katello).

ares committed:
Fixes #19742 - cache hosts for fact values

When we generate fact values hash we always loaded a host for a given
fact. This in combination with taxonomies queryies that are
automatically generated caused roughly 6N+1 issue. We can preload all
hosts that are required and use this in-memory cache for generating the
hash.

ares committed:
Fixes #20708 - fix templates auditing

Templates were audited under parent class. This keeps separate audits
per template type and allows searching by provisioning_template and
partition_table names. It also fixes broken associations in
taxable_taxonomies table.

mmoll committed:
Refs #19707 - test searching domains with locs

bastilian committed:
Fixes #20800 - Comply with Patternfly login page recommendations

* Move alerts inline above form
* Make submit button not show a "loading" state

ares committed:
Fixes #20660 - improve proxy features list

Commit:
Fixes #20897 - change description for 'default' to help tooltip

dhlavac committed:
Fixes #15286 - New API for auth sources

dhlavac committed:
Fixes #15286 - New API for auth sources
